doc/source/basics.rst

```diff
@@ -1154,21 +1154,46 @@ any of its three axes.
 Iteration
 ---------
 
-Because Series is array-like, basic iteration produces the values. Other data
-structures follow the dict-like convention of iterating over the "keys" of the
-objects. In short:
+The behavior of basic iteration over pandas objects depends on the type.
+When iterating over a Series, it is regarded as array-like, and basic iteration
+produces the values. Other data structures, like DataFrame and Panel,
+follow the dict-like convention of iterating over the "keys" of the
+objects.
 
-  * **Series**: values
-  * **DataFrame**: column labels
-  * **Panel**: item labels
+In short, basic iteration (``for i in object``) produces:
 
-Thus, for example:
+* **Series**: values
+* **DataFrame**: column labels
+* **Panel**: item labels
+
+Thus, for example, iterating over a DataFrame gives you the column names:
 
 .. ipython::
 
-   In [0]: for col in df:
-      ...:     print(col)
-      ...:
+    df = pd.DataFrame({'col1' : np.random.randn(3), 'col2' : np.random.randn(3)},
+                      index=['a', 'b', 'c'])
+
+    In [0]: for col in df:
+       ...:     print(col)
+       ...:
+
+Pandas objects also have the dict-like :meth:`~DataFrame.iteritems` method to
+iterate over the (key, value) pairs.
+
+To iterate over the rows of a DataFrame, you can use the following methods:
+
+* :meth:`~DataFrame.iterrows`: Iterate over the rows of a DataFrame as (index, Series) pairs.
+  This converts the rows to Series objects, which can change the dtypes and has some
+  performance implications.
+* :meth:`~DataFrame.itertuples`: Iterate over the rows of a DataFrame as tuples of the values.
+  This is a lot faster as :meth:`~DataFrame.iterrows`, and is in most cases preferable to
+  use to iterate over the values of a DataFrame.
+
+.. warning::
+
+  In many cases, iterating manually over the rows is not needed. Iterating with
+  pandas objects is generally slow, and in most cases vectorized solutions
+  exist.
 
```

Comment:
you also need a big warning that you cannot SET columns and expect it to work (as its operating on a copy). This is also a big source of confusion. 

Further you can mention ``.apply``.

I would highlite the performance issues WAY more here. Encourage vectorization first.